[#5696] Add JCC api calls (2)

vaszig opened the pull request:
Closes #5696 partly

**Changes**

- [x] `get_required_customer_fields`
- [x] `create_appointment`
- [x] `delete_appointment`
- [x] `get_appointment_details`

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -0,0 +1,155 @@
+        self, product_uuids: Sequence[str]
+    ) -> CustomerFields:
+        """Retrieve all the required customer field names based on the products UUIDs."""
+        params = {"selectedActivityId": product_uuids}
```
TIL you can just pass it as a list and it will add a parameter for each item. I did it manually for my API calls :sweat_smile:

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/constants.py:
```diff
@@ -1,5 +1,295 @@
+
+
+def get_component(component_type: CustomerFields, required: bool):
+    FIELD_TO_FORMIO_COMPONENT: dict[CustomerFields, Component] = {
```
Perhaps define `FIELD_TO_FORMIO_COMPONENT` separately and use `get_component` to set the required field
```python
def get_component(component_type: CustomerFields, required: bool):
    component = deepcopy(FIELD_TO_FORMIO_COMPONENT[component_type])
    component["validate"]["required"] = required
    return component
```
Or use a match statement? Seems a bit wasteful to define the whole dict every time a single component is requested

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/constants.py:
```diff
@@ -1,5 +1,295 @@
+            "values": [  # pyright: ignore[reportAssignmentType]
+                {"value": "M", "label": _("Male")},
+                {"value": "F", "label": _("Female")},
+                {"value": "N", "label": _("Not specified")},
```
Shouldn't the values here represent the options we need to send to the endpoint (0, 1, or 2)? Or is there a custom conversion somewhere else?

Also, I think "Other" or "Prefer not to say" is a better label, as it should be an answer to the question for the user "What is your gender?"

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -97,7 +96,74 @@ def get_required_customer_fields(
+            # required ones (required fields in the form).
+            # 0 = visible
+            # 1 = hidden
+            # 2 = required
```
I think adding an enum/constants for these values would improve readability of the statements below :)

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -97,7 +96,74 @@ def get_required_customer_fields(
+            # field should be 'required.'
+            default_number_field = []
+            default_name_or_initials = []
+            if required_fields.get("isAnyPhoneNumberRequired"):
```
What if `isAnyPhoneNumberRequired` is true, and both fields are set to visible? We should mark one of them as required then right?

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -97,7 +96,74 @@ def get_required_customer_fields(
+                initials = required_fields.get("initials")
+                first_name = required_fields.get("firstName")
+                if (not initials or initials == 1) and (
+                    not first_name or first_name == 1
```
Could you add a test what happens when both fields are marked as visible (0), but `areFirstNameOrInitialsRequired` is set to `True`? It seems like we do set a `default_name_or_initials` because of the `not initials` and `not first_name`, but I'm afraid it gets added twice at the end as all fields from `field_names` are added to the list as well.

I think being explicit here in checking the value of the fields might help with readability as well: `initials in (0, 1)`, or using an enum `initials in (state.visible, state.hidden)`

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -97,7 +96,74 @@ def get_required_customer_fields(
+                + [
+                    get_component(CustomerFields(field), required)
+                    for field, required in field_names.items()
+                ]
```
Is this also the order in which the fields are shown in the frontend? If so, perhaps move the default number field to below the default and last name? I know we can't really control the order returned by JCC anyway, but I guess it's the best we can do :shrug:

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/types.py:
```diff
@@ -0,0 +1,195 @@
+class Revision(TypedDict):
+    id: str
+    creationDateTime: str  # ISO 8601 format
+    creatorAuthenticationUserId: str | None
```
Whenever a field has "[type] or null" in the documentation, I think we should treat it as "[type] or missing", which was confirmed by JCC when we asked. Also in my own VCR casssettes, I didn't find a single instance where a field was `null`

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/types.py:
```diff
@@ -0,0 +1,195 @@
+class Customer(TypedDict):
+    id: str | None
+    isMainCustomer: bool
+    gender: FIELD_TYPE
```
I know it's the same values, but could you use a separate literal for this? It has a different meaning than the one for the customer fields

viktorvanwijk reviewed (commented):
Some thoughts and suggestions :)

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -0,0 +1,155 @@
+        )
+
+        params = []
+        for id_, amount in zip(activity_ids, amounts, strict=False):
```
OK, tested it on my endpoint, and it does work. Though maybe better to keep it like this to match the example in the documentation?

vaszig committed:
[#5696] Implemented JCC API calls

This commit has the necessary code for the following calls:
- get_required_customer_fields
- create_appointment
- delete_appointment
- get_appointment_details

A lot of tests use mocks because at this time JCC does not provide the
useful test data to test them properly. A lot of realistic scenarios are
missing from the test instance so VCR testing cannot be used for all our
test suite.

viktorvanwijk commented on src/openforms/appointments/api/serializers.py:
```diff
@@ -282,7 +285,39 @@ def validate(self, attrs: dict) -> dict:
+                if group["type"] != "require_one_of":
+                    continue
+
+                contact_details_meta = deepcopy(contact_details_meta)
```
This copy is not necessary when you set it to the same variable, right?

viktorvanwijk commented on src/openforms/appointments/base.py:
```diff
@@ -73,6 +74,22 @@ def __str__(self):
+class RequireOneOfRule:
+    enabled: bool
+    fields: tuple[str, ...]
+    labels: tuple[str, ...]
```
I feel like this should be part of the JCC plugin specifically, because it's only used as a helper dataclass there. Nothing in the api for `get_required_customer_fields` dictates that this should be used

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -162,6 +171,93 @@ def get_available_times_for_appointment(
+        """Retrieve the duration (in minutes) for an appointment."""
+        assert len(activity_ids) == len(amounts), (
+            "Number of activity IDs and amounts must be equal"
+        )
```
You didn't want to combine them like I did in `get_available_times_for_appointment`?

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/constants.py:
```diff
@@ -1,5 +1,296 @@
+        "values": [  # pyright: ignore[reportAssignmentType]
+            {"value": GenderType.other, "label": _("Other")},
+            {"value": GenderType.male, "label": _("Male")},
+            {"value": GenderType.female, "label": _("Female")},
```
```suggestion
            {"value": GenderType.male.value, "label": GenderType.male.label},
            {"value": GenderType.female.value, "label": GenderType.female.label},
            {"value": GenderType.other.value, "label": GenderType.other.label},
```
Put other at the end maybe?

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -157,12 +177,144 @@ def get_times(
+    ) -> dict[str, bool]:
+        """
+        Based on the retrieved fields returns a dict with the fields and whether they are
+        required or not.
```
```suggestion
        Based on the retrieved fields returns a dict with the fields that should be rendered and whether they are
        required or not.
```

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/plugin.py:
```diff
@@ -157,12 +177,144 @@ def get_times(
+            self._get_default_component(
+                required_fields.get("areFirstNameOrInitialsRequired"),
+                required_fields.get("initials"),
+                required_fields.get("firstName"),
```
These values do not match the type hints of `self._get_default_components` as they can be `None` as well. Suggestion to update the type hints or add default values

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/tests/test_plugin.py:
```diff
@@ -29,7 +32,39 @@ def _scrub_access_token(response):
+       - endpoints must work as expected
+       - data is still valid
+    #. Delete the VCR cassettes
+    #. Run the tests
```
```suggestion
    #. Delete the VCR cassettes
    #. Update the ``RECORDING_DATETIME`` class attribute to the date of recording
    #. Run the tests
```

viktorvanwijk reviewed (commented):
General flow looks OK, just a few minor things :)

One thought I had is to put all of the mocked requested into a separate test class, which will avoid a lot of VCR cassettes for the token requests. But we can leave that for later. Regarding the mocks themselves, I think you've covered all the relevant cases now. I would be nice of course if we could get this data from JCC, but I'm really not sure if we should spend the time on it to be honest. If you want to send the email, feel free.

Could you do me a favor and process the feedback in a separate commit :innocent:

vaszig commented:
> General flow looks OK, just a few minor things :)
> 
> One thought I had is to put all of the mocked requested into a separate test class, which will avoid a lot of VCR cassettes for the token requests. But we can leave that for later.
> 
> Could you do me a favor and process the feedback in a separate commit 😇

My apologies, I am used to other people reviewing the processed feedback by clicking the "compare" button. I will do that in a separate commit though!

vaszig committed:
[#5696] Processed feedback

vaszig commented on src/openforms/appointments/contrib/jcc_rest/types.py:
```diff
@@ -107,25 +96,25 @@ class Location(BaseLocation):
-    gender: GENDER_TYPE
-    firstName: FIELD_STATE
-    initials: FIELD_STATE
+    gender: GenderType
```
Stupid..I had it correct and last minute I saw it and made it wrong again. I am fixing this.
